matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/network/token/HomeserverAccessTokenProvider.kt

```diff
@@ -16,13 +16,99 @@
 
 package org.matrix.android.sdk.internal.network.token
 
+import kotlinx.coroutines.runBlocking
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
+import org.matrix.android.sdk.api.auth.AuthenticationService
+import org.matrix.android.sdk.api.auth.data.RefreshResult
+import org.matrix.android.sdk.api.failure.Failure
+import org.matrix.android.sdk.api.failure.MatrixError.Companion.M_FORBIDDEN
+import org.matrix.android.sdk.api.failure.MatrixError.Companion.M_UNKNOWN_TOKEN
 import org.matrix.android.sdk.internal.auth.SessionParamsStore
 import org.matrix.android.sdk.internal.di.SessionId
+import org.matrix.android.sdk.internal.network.GlobalErrorReceiver
+import org.matrix.android.sdk.internal.network.toFailure
+import retrofit2.HttpException
+import timber.log.Timber
 import javax.inject.Inject
 
 internal class HomeserverAccessTokenProvider @Inject constructor(
         @SessionId private val sessionId: String,
-        private val sessionParamsStore: SessionParamsStore
+        private val sessionParamsStore: SessionParamsStore,
+        private val authenticationService: AuthenticationService,
+        private val globalErrorReceiver: GlobalErrorReceiver
 ) : AccessTokenProvider {
-    override fun getToken() = sessionParamsStore.get(sessionId)?.credentials?.accessToken
+
+    companion object {
+        private val mutex = Mutex()
+    }
+
+    object Constants {
+        /**
+        The time interval before the access token expires that we will start trying to refresh the token.
+        This avoids us having to block other users requests while the token refreshes.
+        Choosing a value larger than DEFAULT_DELAY_MILLIS + DEFAULT_LONG_POOL_TIMEOUT_SECONDS guarantees we will at least have attempted it before expiry.
+         */
+        const val PREEMPT_REFRESH_EXPIRATION_INTERVAL = 60000
```

Comment:
could be declared in the `companion object` above (no need for a new `Constants` object.)